### PR TITLE
Implement Update for CleaningEvent

### LINKED ISSUE #125: Add Update endpoint for the CleaningEventController
## Description

Implement a new new `Update` method for the `CleaningEventController`, which should take the `CleaningEventUpdateRequest` as a `Patch` body.

## Steps to take

1. Take a look at the update implementation found in the `PollutedLocationController`. You can use it as a reference. 
2. Create the needed public models, its validators, adjust the cleaning event mapping profile.
3. Write unit tests to cover the implemented functionality in the controller, validator, mapper and other parts if modified.

## Acceptance criteria

You can update an instance of CleaningEvent using a patch request in swagger, with the endpoint /Update/.
The implementation is covered with Unit tests.

### PR DESCRIPTION
**Pull request by BaVilius**
Created an Update Method and wrote tests.
<hr>
Key question: Should we let change the PollutedLocation while updating a CleaningEvent? (**UPDATED** : We shouldn't)

Closes #125

### CODE REVIEW

**Comment by BenasB** [comment]
Closes #125

**Comment by BenasB** [comment]
About `ParentExistsByPropertyAsync` see https://github.com/vu-vibedosa/apsitvarkom-be/pull/131#discussion_r1038964617
(using PollutedLocation repository instead of `ParentExistsByPropertyAsync`)

> Should we let change the PollutedLocation while updating a CleaningEvent?

IMO, no. We can't change the location in `PollutedLocation` and expect the user to delete the polluted location and create another one. I think this should behave in the same way: if a user messes up and creates a cleaning event with the wrong polluted location, they should delete it and create a new one with the correct polluted location id

**Comment by BaVilius** [comment]
> About `ParentExistsByPropertyAsync` see [#131 (comment)](https://github.com/vu-vibedosa/apsitvarkom-be/pull/131#discussion_r1038964617) (using PollutedLocation repository instead of `ParentExistsByPropertyAsync`)
> 
> > Should we let change the PollutedLocation while updating a CleaningEvent?
> 
> IMO, no. We can't change the location in `PollutedLocation` and expect the user to delete the polluted location and create another one. I think this should behave in the same way: if a user messes up and creates a cleaning event with the wrong polluted location, they should delete it and create a new one with the correct polluted location id

So CleaningEventUpdateRequest should only contain Id, StartTime and Notes?

**Comment by BenasB** [comment]
> So CleaningEventUpdateRequest should only contain Id, StartTime and Notes?

Yes, that's how I see it: Id to identify which cleaning event to update and the other fields for actually updating the information.

Unless someone thinks we would benefit from allowing to change the polluted location as well, of course, open for discussion

P.S. I did not have the time to look into the PR changes now, but just wanted to comment on the PR description

**Comment by BaVilius** [comment]
> P.S. I did not have the time to look into the PR changes now, but just wanted to comment on the PR description

That's good, I need to do some changes before asking for a review then

File: Apsitvarkom.Api/Controllers/CleaningEventController.cs [hunk 1]
```diff
@@ -11,18 +11,24 @@ namespace Apsitvarkom.Api.Controllers;
 [Route("/api/[controller]")]
 public class CleaningEventController : ControllerBase
 {
-    private readonly IRepository<CleaningEvent> _repository;
+    private readonly IRepository<CleaningEvent> _cleaningEventRepository;
+    private readonly IPollutedLocationRepository _pollutedLocationRepository;
```
Comment (6ar8nas) [inline Apsitvarkom.Api/Controllers/CleaningEventController.cs]: this PR doesn't require PollutedLocationRepository to be introduced - I believe it is a left-over after the initial implementation. revert the _cleaningEventRepository name as well.

File: Apsitvarkom.Models/Public/CleaningEventUpdateRequest.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    public string? Notes { get; set; }
+}
+
+public class CleaningEventUpdateRequestValidator : AbstractValidator<CleaningEventUpdateRequest>
+{
+    public CleaningEventUpdateRequestValidator()
+    {
+        RuleFor(l => l.Id).NotNull();
+        RuleFor(l => l.StartTime).NotNull().GreaterThan(DateTime.UtcNow);
```
Comment (6ar8nas) [inline Apsitvarkom.Models/Public/CleaningEventUpdateRequest.cs]: we should be able to have start time as null - let's take an update request, where we only want to update notes as an example.

File: Apsitvarkom.Models/Mapping/CleaningEventProfile.cs [hunk 3]
```diff
@@ -19,6 +19,12 @@ private void MapRequests()
     {
         CreateMap<ObjectIdentifyRequest, CleaningEvent>(MemberList.None)
             .ForMember(x => x.Id, opt => opt.MapFrom(x => x.Id));
+        CreateMap<CleaningEventUpdateRequest, CleaningEvent>()
+            .ForMember(x => x.Id, opt => opt.PreCondition(src => src.Id is not null))
```
Comment (6ar8nas) [inline Apsitvarkom.Models/Mapping/CleaningEventProfile.cs]: we already check this precondition in the validator. we are sure, that the `CleaningEventUpdateRequest` has an id, so I don't think is really necessary.
However, I also don't think this is a mistake, so whatever you decide is good [it's just that I've noticed the `PollutedLocationUpdateRequest` doesn't have this, while having stuff match is preferred]

**Review by 6ar8nas — CHANGES_REQUESTED** [review]
some minor changes. didn't have the time to look at the tests yet.

File: Apsitvarkom.UnitTests/Api/Controllers/CleaningEventControllerTests.cs [hunk 4]
```diff
@@ -8,6 +8,7 @@
 using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Mvc;
 using Moq;
+using NUnit.Framework.Constraints;
```
Comment (6ar8nas) [inline Apsitvarkom.UnitTests/Api/Controllers/CleaningEventControllerTests.cs]:
```suggestion
```

Comment (6ar8nas) [reply Apsitvarkom.UnitTests/Api/Controllers/CleaningEventControllerTests.cs]: seems like an unused reference

File: Apsitvarkom.UnitTests/Api/Controllers/CleaningEventControllerTests.cs [hunk 5]
```diff
@@ -214,6 +217,111 @@ public async Task GetById_RepositoryThrows_Status500InternalServerErrorReturned(
     }
     #endregion
 
+    #region Update tests
+    public async Task Update_NullIdEntered_ValidationResultsInBadRequestResponseReturned()
```
Comment (6ar8nas) [inline Apsitvarkom.UnitTests/Api/Controllers/CleaningEventControllerTests.cs:221]: `[Test]` attribute is missing - we don't ever test this method with the current setup.

File: Apsitvarkom.UnitTests/Api/Controllers/CleaningEventControllerTests.cs [hunk 6]
```diff
@@ -214,6 +217,111 @@ public async Task GetById_RepositoryThrows_Status500InternalServerErrorReturned(
...
+        var resultEvent = result.Value as CleaningEventResponse;
+        Assert.That(resultEvent, Is.Not.Null);
+        Assert.Multiple(() =>
+        {
+            Assert.That(resultEvent.Id, Is.EqualTo(cleaningEvent.Id));
+            Assert.That(resultEvent.StartTime, Is.EqualTo(cleaningEvent.StartTime));
+            Assert.That(resultEvent.Notes, Is.EqualTo(cleaningEvent.Notes));
+        });
+    }
+
```
Comment (6ar8nas) [inline Apsitvarkom.UnitTests/Api/Controllers/CleaningEventControllerTests.cs]: Test covering a case where `UpdateAsync` throws is missing.

File: Apsitvarkom.UnitTests/Models/Validation/CleaningEventUpdateValidatorTests.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
+﻿using Apsitvarkom.Models.Public;
+using FluentValidation;
+
+namespace Apsitvarkom.UnitTests.Models.Validation;
+
+public class CleaningEventUpdateValidatorTests
```
Comment (6ar8nas) [inline Apsitvarkom.UnitTests/Models/Validation/CleaningEventUpdateValidatorTests.cs]: looking at our code we typically have the name `'PublicModel'ValidationTests` here

**Review by 6ar8nas — CHANGES_REQUESTED** [review]

**Review by 6ar8nas — APPROVED** [review]
looks good.

**Pull request merged**